### PR TITLE
Fix EIP712 encoding

### PR DESCRIPTION
**Pull request by Uxio0**
- Fix empty arrays
- Fix arrays with repeating elements

### CODE REVIEW

**Review by mikhailxyz — APPROVED** [review]
🔝

**Comment by Uxio0** [comment]
Using the following node script to validate the hashes:

```js
const { ethers } = require('ethers')

const typedData = {
    "types": {
        "Nested": [
            {
                "name": "nestedString",
                "type": "string"
            },
            {
                "name": "nestedAddress",
                "type": "address"
            },
            {
                "name": "nestedUint256",
                "type": "uint256"
            },
            {
                "name": "nestedUint32",
                "type": "uint32"
            },
            {
                "name": "nestedBytes32",
                "type": "bytes32"
            },
            {
                "name": "nestedBoolean",
                "type": "bool"
            }
        ],
        "Example": [
            {
                "name": "testString",
                "type": "string"
            },
            {
                "name": "testAddress",
                "type": "address"
            },
            {
                "name": "testUint256",
                "type": "uint256"
            },
            {
                "name": "testUint32",
                "type": "uint32"
            },
            {
                "name": "testBytes32",
                "type": "bytes32"
            },
            {
                "name": "testBoolean",
                "type": "bool"
            },
            {
                "name": "testNested",
                "type": "Nested"
            },
            {
                "name": "testNestedArray",
                "type": "Nested[]"
            }
        ],
        "EIP712Domain": [
            {
                "name": "name",
                "type": "string"
            },
            {
                "name": "version",
                "type": "string"
            },
            {
                "name": "chainId",
                "type": "uint256"
            },
            {
                "name": "verifyingContract",
                "type": "address"
            }
        ]
    },
    "domain": {
        "name": "EIP-1271 Example DApp",
        "version": "1.0",
        "chainId": "5",
        "verifyingContract": "0xaecdfd3a19f777f0c03e6bf99aafb59937d6467b"
    },
    "primaryType": "Example",
    "message": {
        "testString": "Hello Deeeeeemo",
        "testAddress": "0xaecdfd3a19f777f0c03e6bf99aafb59937d6467b",
        "testUint256": "115792089237316195423570985008687907853269984665640564039457584007908834671663",
        "testUint32": "123",
        "testBytes32": "0x00000000000000000000000000000000000000000000000000000000deadbeef",
        "testBoolean": true,
        "testNested": {
            "nestedString": "Hello Deeeeeemo",
            "nestedAddress": "0x0000000000000000000000000000000000000002",
            "nestedUint256": "0",
            "nestedUint32": "1",
            "nestedBytes32": "0x000000000000000000000000000000000000000000000000000000000000da7a",
            "nestedBoolean": false
        },
        "testNestedArray": [
            {
                "nestedString": "Hello Deeeeeemo",
                "nestedAddress": "0x0000000000000000000000000000000000000002",
                "nestedUint256": "0",
                "nestedUint32": "1",
                "nestedBytes32": "0x000000000000000000000000000000000000000000000000000000000000da7a",
                "nestedBoolean": false
            },
            {
                "nestedString": "Hello Deeeeeemo",
                "nestedAddress": "0x0000000000000000000000000000000000000002",
                "nestedUint256": "0",
                "nestedUint32": "1",
                "nestedBytes32": "0x000000000000000000000000000000000000000000000000000000000000da7a",
                "nestedBoolean": false
            }
        ]
    }
}

const { EIP712Domain: _, ...types } = typedData.types
console.log(ethers.utils._TypedDataEncoder.hash(typedData.domain, types, typedData.message))
```

File: gnosis/eth/eip712/__init__.py [hunk 1]
```diff
@@ -75,15 +76,16 @@ def _encode_field(name, typ, value):
...
-            h = fast_keccak(
-                encode_abi(
-                    list(type_value_pairs.keys()), list(type_value_pairs.values())
-                )
-            )
+            # Array type
+            parsed_type = typ[: typ.rindex("[")]
+            type_value_pairs = [_encode_field(name, parsed_type, v) for v in value]
+            # Empty array
+            if value:
```
Comment (moisses89) [inline gnosis/eth/eip712/__init__.py:83]: I suggest to check if value exists before to get the `type_value_pairs` you can do something like:
```
     if value:
          parsed_type = typ[: typ.rindex("[")]
          type_value_pairs = [_encode_field(name, parsed_type, v) for v in value]
          data_types, data_hashes = zip(*type_value_pairs)
      else: 
          data_types, data_hashes = [], []
```

**Review by moisses89 — COMMENTED** [review]
Looks good only a suggestion :)

File: gnosis/eth/eip712/__init__.py (hunk 1 above)
Comment (Uxio0) [reply gnosis/eth/eip712/__init__.py:83]: Good suggestion, more optimal!

Comment (Uxio0) [reply gnosis/eth/eip712/__init__.py:83]: Done

**Review by moisses89 — APPROVED** [review]
:rocket:

**Pull request merged**